Fix cp.app.prefs (Issue #1355)

latenitefilms opened the pull request:
- Switched `hs._asm.cfpreferences` to use the bundle ID instead of the
property list path
- Added a `synchronize` prior to the `get`

latenitefilms committed:
#1355

- Switched `hs._asm.cfpreferences` to use the bundle ID instead of the
property list path
- Added a `synchronize` prior to the `get`

randomeizer commented:
Make sure you treat that this works on sandboxed apps such as TextEdit or Preview.

latenitefilms commented:
@randomeizer - Will do! Don't think I'm going to have much free time this week as we're shooting, but will try and have a play on the weekend.

FYI: `hs._asm.cfpreferences.setValue()` has some optional parameters for defining the user name and host name, so that might be why you were having issues with sandboxed apps? I haven't looked into it properly though yet.

Here's the relevant code:

```
static int cfpreferences_setAppValue(lua_State *L) {
    LuaSkin *skin = [LuaSkin shared] ;
    [skin checkArgs:LS_TSTRING, LS_TANY, LS_TSTRING | LS_TOPTIONAL, LS_TBOOLEAN | LS_TOPTIONAL, LS_TBOOLEAN | LS_TOPTIONAL, LS_TBREAK] ;

    BOOL              useSimplified  = YES ;
    NSString          *key           = [skin toNSObjectAtIndex:1] ;
    CFPropertyListRef value          = (lua_type(L, 2) != LUA_TNIL) ? luaObjectToCFPropertyListRef(L, 2) : NULL ;
    NSString          *applicationID = (lua_type(L, 3) == LUA_TSTRING) ? [skin toNSObjectAtIndex:3]
                                                                       : (__bridge NSString *)kCFPreferencesCurrentApplication ;
    CFStringRef userName = kCFPreferencesCurrentUser ;
    CFStringRef hostName = kCFPreferencesAnyHost ;

    switch(lua_gettop(L)) {
        case 3:
            if (lua_type(L, 3) == LUA_TBOOLEAN) {
                useSimplified = NO ;
                userName = lua_toboolean(L, 3) ? kCFPreferencesAnyUser : kCFPreferencesCurrentUser ;
            }
            break ;
        case 4:
            useSimplified = NO ;
            if (lua_type(L, 3) == LUA_TBOOLEAN) {
                userName = lua_toboolean(L, 3) ? kCFPreferencesAnyUser : kCFPreferencesCurrentUser ;
                hostName = lua_toboolean(L, 4) ? kCFPreferencesAnyHost : kCFPreferencesCurrentHost ;
            } else { // lua_type(L, 2) == LUA_TSTRING
                userName = lua_toboolean(L, 4) ? kCFPreferencesAnyUser : kCFPreferencesCurrentUser ;
            }
            break ;
        case 5:
            useSimplified = NO ;
            userName = lua_toboolean(L, 4) ? kCFPreferencesAnyUser : kCFPreferencesCurrentUser ;
            hostName = lua_toboolean(L, 5) ? kCFPreferencesAnyHost : kCFPreferencesCurrentHost ;
            break ;
    }

    if (value == NULL && lua_type(L, 2) != LUA_TNIL) {
        return luaL_argerror(L, 2, "invalid value specified; details in log") ;
    }

    if (CFEqual(userName, kCFPreferencesAnyUser)) {
        [skin logWarn:[NSString stringWithFormat:@"%s.setValue with anyUser set to true requires administrator privileges which is not supported by Hammerspoon; this is expected to fail.", USERDATA_TAG]] ;
    }

    if (useSimplified) {
        CFPreferencesSetAppValue((__bridge CFStringRef)key, value, (__bridge CFStringRef)applicationID) ;
    } else {
        CFPreferencesSetValue((__bridge CFStringRef)key, value, (__bridge CFStringRef)applicationID, userName, hostName) ;
    }
    return 0 ;
}
```

latenitefilms committed:
#1355

- Tweaked the HUD
- Fixed bug in `cp.app.prefs.get()` that would prevent the function
from returning `false`

latenitefilms committed:
#1355

- Cleaned up code
